mantidproject/mantid
# Add identifier for output workspaces in Elwin Interface

### Description of work
This PR creates a new widget that can be embedded in inelastic interfaces to add a label to the name of the output workspace. The label is typically added before the default workspace prefix( for example, _red or _sqw), unless there is not prefix on the interface, in which case is added at the end of the default output name (although the algorithm could add another prefix on the final workspace generation). When the label is changed, the widget will generate what would be the final workspace name after running the interface algorithm and check if this name is already in use in the ADS. Then updating a colored label with an appropriate warning message so that the user can know whether data will be overridden, but it does not prevent the data to be overridden if the user still runs the interface algorithm. 



Fixes #37201 




#### Further detail of work

Initially, I added the widget as a single class containing all the functionality. On a second revision, I have changed it to contain a presenter and a view to make it more organized and also to add tests. 

### To test:
- Open Elwin interface. 
- Make sure the restrict workspace by extension... option is checked on settings.
- Add a sample file to the interface: `irs26176_graphite002_red`
- The label widget sits just below the `Run` widget. You will see that next to the `Current Output Name` text, a label will update with the full expected output name for the current data file. The warning label just before this label is turned on green color with a text indicating that with the current output name, there won't be any overriding workspaces on the ads. 
- There is an edit with the placeholder text `NoLabel`. Click on the edit and write `label`
- The current output name should update to: `irs26176_graphite002_label_red_elwin_eq`, still the warning label on green as this workspace name is not used on the ADS. 
- Click on Run. 
- The warning label will immediately turn to red, indicating that the current label is in used on the ADS. 
- Change the label to `label2`, the warning label will be green indicating the output name is not overridden. 
- On settings, untick the restrict by suffix option. 
- When changing the label, you should see that the label appears after the workspace suffix: `irs26176_graphite002_red_label...` 







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
